nexus/db-queries/src/authz/omicron.polar

```diff
@@ -111,6 +111,11 @@ resource Fleet {
 	"modify" if "admin";
 }
 
+# For fleets specifically, roles can be conferred by roles on the user's Silo.
+has_role(actor: AuthenticatedActor, role: String, _: Fleet) if
+	silo_role in actor.confers_fleet_role(role) and
+	has_role(actor, silo_role, actor.silo.unwrap());
```

Comment:
Yes!  This is a surprisingly deep question and I got this wrong twice while implementing this.

The `has_role()` rule in the Polar file is by definition the canonical way to determine if an actor has a role on a resource.  There can be many such rules in the Polar file.  We usually think about this one that calls right into the Rust `has_role` method:
https://github.com/oxidecomputer/omicron/blob/e37b943ac895384717294ee8f45e4c8152ef435e/nexus/db-queries/src/authz/omicron.polar#L25-L26

This method would probably be better called `has_directly_assigned_role()` because that's really what it checks.  This is _one_ way that an actor can have a role but it's not the only one.

We define other rules in the Polar file, like these two:
https://github.com/oxidecomputer/omicron/blob/e37b943ac895384717294ee8f45e4c8152ef435e/nexus/db-queries/src/authz/omicron.polar#L550C3-L553

We also implicitly define other `has_role` rules with stuff like this:
https://github.com/oxidecomputer/omicron/blob/e37b943ac895384717294ee8f45e4c8152ef435e/nexus/db-queries/src/authz/omicron.polar#L128-L130

In earlier commits, I tried to implement this behavior in Rust in the implementation of one of the Polar methods.  The code would say something like: "if the Silo is configured such that the 'viewer' role on the Silo confers the 'viewer' role on the Fleet, then if I find the 'viewer' role on the Silo in the user's roleset, then I will _add_ the 'viewer' role on the Fleet to the user's roleset."  But suppose the actor has the 'admin' role on the Silo.  'viewer' will not appear in the roleset.   (The roleset is just the set of directly assigned roles.)  That's wrong -- someone who has "admin" on the Silo _also_ has "viewer" on the Silo, so they should also get "viewer" on the Fleet.  But in order to know this, you have to evaluate the Polar expression `has_role()`.  You can't figure this out by looking at just the directly attached roles.

(Aside: I don't usually curate my commits on GitHub PRs so I'm sorry if you wasted time going through my stream of consciousness of commits.)

Reply:
Ah, that makes sense, thanks!